### ISSUE TITLE
Asynchronous API

### ISSUE THREAD

**Issue by ojhaujjwal** [thephpleague/flysystem#198]
With the evolution of [react](https://github.com/reactphp/react), it's time now to think about creating asynchronous API.

**Comment by Webysther** [comment]
@frankdejonge After 3 years something change? Consider only for local adapter?

**Comment by frankdejonge** [comment]
@Webysther unfortunately PHP hasn't changed that much.

**Comment by esroyo** [comment]
> After some good talks with @WyriHaximus I've come to the conclusion that this, with this current state of async in PHP, is not viable. Pretty much all the adapters would have to be rewritten in the async mindset, this is as much out of scope as it can be for Flysystem.

Hi @frankdejonge (cc @WyriHaximus)

How about the possibility of _just_ standardizing an asynchronous API and let the ecosystem grow naturally? Flysystem does a beautiful job by providing a unified interface. We could augment that interface with async equivalent methods. Thus in the same way Guzzle has `get` \ `getAsync`, `post` \ `postAsync` and so on, Flysystem could have `$filesystem->write()` and `$filesystem->writeAsync()`. All continues to be the same, except results are wrapped in a Promise.

I made a prototype to "visualize" the idea: https://git.sr.ht/~esroyo/php-flysystem-async

Interfaces:
https://git.sr.ht/~esroyo/php-flysystem-async/tree/master/src/FilesystemAsyncInterface.php
https://git.sr.ht/~esroyo/php-flysystem-async/tree/master/src/AdapterAsyncInterface.php

When an adapter doesn't implement the AdapterAsyncInterface, async methods can be generated by just wrapping the usual sync result in a Promise. Or the other way around, an async adapter can be made synchronous by blocking until the promise resolves (like [here](https://gitlab.expoteca.com/esroyo/php-flysystem-async/blob/master/src/Adapter/AsyncToSyncTrait.php)). Anyway that could be up to the adapter.

There would be *a lot* to discuss, but the final usage example is very close to what @ojhaujjwal was asking for:
https://git.sr.ht/~esroyo/flysystem-async-example/tree/master/item/example.php

How does that sound?

**Comment by devnix** [comment]
@esroyo sounds sweet. Should we open a separate ticket to write an RFC about this? The Filesystem component of ReactPHP looks gorgeous, and seems like there is an FTP component in the kitchen https://twitter.com/another_clue/status/1391768024605241349

I would love to humbly contribute to the project or the adapters. For me, the hardest part would be to have a consensus on the final API.

**Comment by esroyo** [comment]
@devnix sorry for the late answer. I'm not using so much PHP this days, so I really lost interest.
I hope PHP 8.1 Fibers stimulates interest on async a bit.

**Comment by RoofTurkey** [comment]
@frankdejonge how about an extra method in the interface for writing multiple files in one call. `writeMultiple(array $files)`, where `$files` could be a `File` model containing a path, content and config. This way an adapter could implement this function in whatever way makes sense for that adapter (synchronously or asynchronously). 

For example, we use Google Cloud Storage and the performance (synchronously) is simply too poor in combination with FlySystem. We now use the `google/cloud-storage` directly with `uploadAsync()` (instead of `upload()`) and the performance increase is enormous, where saving (30) files now only takes as long as one file would.

At the same time, the call will be almost as simple as it is now. Without the need for promises, even though that does mean you wouldn't have the freedom to use asynchronous saving directly. The caller would always have to prepare all the files before the first one can start being saved.

A basic implementation for adapters would simply be:
```
/** @param File[] $files */
public function writeMultiple(array $files): void
{
    foreach($files as $file) {
        $this->write($file->getPath(), $file->getContents(), $file->getConfig());
    }
}
```

Anyways, thank you (and the other contributors) for your great work on FlySystem. Hopefully my suggestion can be used to make it even better, I'd be happy to help implement it

**Comment by swiffer** [comment]
Definitly interested in Async / Batch / Pool operations running in parallel (as working with a lot of files over S3 is slow otherwise). Would be awesome to see support for this.

I temporarly switched to the AWS PHP SDK as it's supporting this via Command Pools and it's a real game changer. Copying 80.000 files took 5 minutes with a concurrency level of 50.

https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_commands.html#command-pool

**Comment by WyriHaximus** [comment]
As long as the underlying I/O can be replaced through adapter configuration, no other changes should be required to flysystem. (In theory.)
